storm-core/src/jvm/backtype/storm/blobstore/BlobStoreAclHandler.java

```diff
@@ -0,0 +1,387 @@
+    ret.append(":");
+    if (ac.is_set_name()) {
+      ret.append(ac.get_name());
+    }
+    ret.append(":");
+    ret.append(accessToString(ac.get_access()));
+    return ret.toString();
+  }
+
+  public static void validateSettableACLs(String key, List<AccessControl> acls) throws AuthorizationException {
+    Set<String> aclUsers = new HashSet<>();
+    List<String> duplicateUsers = new ArrayList<>();
+    for (AccessControl acl : acls) {
+      String aclUser = acl.get_name();
+      if (aclUser != null && !aclUser.isEmpty() && !aclUsers.add(aclUser)) {
+        LOG.error("'{}' user can't appear more than once in the ACLs", aclUser);
+        duplicateUsers.add(aclUser);
+      }
+    }
+    if (duplicateUsers.size() > 0) {
+      String errorMessage  = "user " + Arrays.toString(duplicateUsers.toArray())
+          + " can't appear more than once in the ACLs for key [" + key +"].";
+      throw new AuthorizationException(errorMessage);
+    }
+  }
+
+  private Set<String> constructUserFromPrincipals(Subject who) {
+    Set<String> user = new HashSet<String>();
+    if (who == null) {
+      LOG.debug("in validate acl who is null");
+    } else {
+      LOG.debug("in validate acl: " + who);
+    }
+    if (who != null) {
+      for (Principal p : who.getPrincipals()) {
+        user.add(_ptol.toLocal(p));
+      }
+    }
+    return user;
+  }
+
+  private boolean isSupervisorOrAdmin(Set<String> user, int mask) {
+    boolean isSupervisor = false;
+    boolean isAdmin = false;
+    for(String u : user) {
+      if (_supervisors.contains(u)) {
+        isSupervisor = true;
+        break;
+      }
+      if (_admins.contains(u)) {
+        isAdmin = true;
+        break;
+      }
+    }
+    if (mask > 0 && !isAdmin) {
+      isSupervisor = (isSupervisor && (mask == 1));
+    }
+    return isSupervisor || isAdmin;
+  }
+
+  private boolean isNimbus(Subject who) {
+    Set<Principal> principals = null;
+    boolean isNimbusInstance = false;
+    if(who != null) {
+      principals = who.getPrincipals();
+      for (Principal principal : principals) {
+        if (principal instanceof NimbusPrincipal) {
+          isNimbusInstance = true;
+        }
+      }
+    }
+    return isNimbusInstance;
+  }
+
+  /**
+   * The user should be able to see the metadata if and only if they have any of READ, WRITE, or ADMIN
+   */
+  public void validateUserCanReadMeta(List<AccessControl> acl, Subject who, String key)
+      throws AuthorizationException {
+    Set<String> user = constructUserFromPrincipals(who);
+    if (isNimbus(who)) {
+      return;
+    }
+    if (isSupervisorOrAdmin(user, -1)) {
+      return;
+    }
+    for (AccessControl ac : acl) {
+      int allowed = getAllowed(ac, user);
+      LOG.debug(" user: {} allowed: {} key: {}", user, allowed, key);
+      if ((allowed & (READ | WRITE | ADMIN)) > 0) {
+        return;
+      }
+    }
+    throw new AuthorizationException(
+            user + " does not have access to " + key);
+  }
```

Comment:
This is mostly duplicate code.  We could just call `validateAnyACL` with the RWA mask.